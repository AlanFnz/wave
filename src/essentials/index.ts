export { Elevation } from './Elevation/Elevation';
export { Spaces } from './Spaces/Spaces';
export { Breakpoints, MediaQueries } from './Breakpoints/Breakpoints';
export { ClassicColors, ModernColors } from './Colors/globalStyles';
