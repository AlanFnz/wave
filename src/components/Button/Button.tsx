import * as React from 'react';
import styled, { css } from 'styled-components';
import { ButtonStyleProps, ResponsiveValue, variant } from 'styled-system';
import { Colors } from '../../essentials';

import { BaseButton, BaseButtonProps } from './BaseButton';

interface ButtonProps extends BaseButtonProps, ButtonStyleProps {
    /**
     * Adjust colors for display on a dark background
     */
    inverted?: boolean;
    /**
     * Define style of the button component, defaults to primary
     */
    variant?: ResponsiveValue<'primary' | 'secondary' | 'danger'>;
}

const variantStyles = variant({
    variants: {
        primary: {
            background: Colors.AUTHENTIC_BLUE_900,
            borderColor: Colors.AUTHENTIC_BLUE_900,
            color: Colors.WHITE,
            fill: Colors.WHITE,
            '&:hover': {
                background: Colors.AUTHENTIC_BLUE_1100,
                borderColor: Colors.AUTHENTIC_BLUE_1100
            },
            '&:disabled': {
                color: Colors.WHITE,
                fill: Colors.WHITE,
                background: Colors.AUTHENTIC_BLUE_200,
                borderColor: Colors.AUTHENTIC_BLUE_200
            }
        },
        secondary: {
            background: Colors.WHITE,
            borderColor: Colors.AUTHENTIC_BLUE_200,
            color: Colors.AUTHENTIC_BLUE_900,
            fill: Colors.AUTHENTIC_BLUE_900,
            '&:hover': {
                background: Colors.AUTHENTIC_BLUE_50
            },
            '&:disabled': {
                color: Colors.AUTHENTIC_BLUE_200,
                fill: Colors.AUTHENTIC_BLUE_200,
                background: Colors.WHITE,
                borderColor: Colors.AUTHENTIC_BLUE_200
            }
        },
        danger: {
            background: Colors.NEGATIVE_ORANGE_900,
            borderColor: Colors.NEGATIVE_ORANGE_900,
            color: Colors.WHITE,
            fill: Colors.WHITE,
            '&:hover': {
                background: Colors.NEGATIVE_ORANGE_1000,
                borderColor: Colors.NEGATIVE_ORANGE_1000
            },
            '&:disabled': {
                color: Colors.WHITE,
                fill: Colors.WHITE,
                background: Colors.AUTHENTIC_BLUE_200,
                borderColor: Colors.AUTHENTIC_BLUE_200
            }
        }
    }
});

const invertedVariantStyles = variant({
    variants: {
        primary: {
            background: Colors.WHITE,
            borderColor: Colors.WHITE,
            color: Colors.ACTION_BLUE_900,
            fill: Colors.ACTION_BLUE_900,
            '&:hover': {
                background: Colors.ACTION_BLUE_900,
                borderColor: Colors.ACTION_BLUE_900,
                color: Colors.WHITE,
                fill: Colors.WHITE
            },
            '&:disabled': {
                background: Colors.AUTHENTIC_BLUE_550,
                borderColor: Colors.AUTHENTIC_BLUE_550,
                color: Colors.AUTHENTIC_BLUE_350,
                fill: Colors.AUTHENTIC_BLUE_350
            }
        },
        secondary: {
            background: 'transparent',
            borderColor: Colors.WHITE,
            color: Colors.WHITE,
            fill: Colors.WHITE,
            '&:hover': {
                background: Colors.WHITE,
                color: Colors.AUTHENTIC_BLUE_900,
                fill: Colors.AUTHENTIC_BLUE_900
            },
            '&:disabled': {
                background: 'transparent',
                borderColor: Colors.AUTHENTIC_BLUE_550,
                color: Colors.AUTHENTIC_BLUE_550,
                fill: Colors.AUTHENTIC_BLUE_550
            }
        },
        danger: {
            background: Colors.NEGATIVE_ORANGE_900,
            borderColor: Colors.NEGATIVE_ORANGE_900,
            color: Colors.WHITE,
            fill: Colors.WHITE,
            '&:hover': {
                background: Colors.NEGATIVE_ORANGE_1000,
                borderColor: Colors.NEGATIVE_ORANGE_1000
            },
            '&:disabled': {
                background: Colors.AUTHENTIC_BLUE_550,
                borderColor: Colors.AUTHENTIC_BLUE_550,
                color: Colors.AUTHENTIC_BLUE_350,
                fill: Colors.AUTHENTIC_BLUE_350
            }
        }
    }
});

const Button: React.FC<ButtonProps> = styled(BaseButton)<ButtonProps>`
    transition: background ease 216ms, border-color ease 216ms;

    ${p => (p.inverted ? invertedVariantStyles : variantStyles)};
`;

Button.defaultProps = {
    size: 'medium',
    variant: 'primary'
};

export { Button, ButtonProps };
