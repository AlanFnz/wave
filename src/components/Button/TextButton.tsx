import * as React from 'react';
import styled from 'styled-components';
import { ResponsiveValue, variant } from 'styled-system';
import { Colors } from '../../essentials';

import { BaseButton, BaseButtonProps } from './BaseButton';

interface TextButtonProps extends BaseButtonProps {
    /**
     * adjust color for display on a dark background
     */
    inverted?: boolean;
    /**
     * Define base colors
     */
    variant?: ResponsiveValue<'default' | 'danger'>;
}

const dangerStyles = {
    color: Colors.NEGATIVE_ORANGE_900,
    fill: Colors.NEGATIVE_ORANGE_900,
    '&:hover': {
        color: Colors.NEGATIVE_ORANGE_1000,
        fill: Colors.NEGATIVE_ORANGE_1000
    }
};

const variantStyles = variant({
    variants: {
        default: {
            color: Colors.ACTION_BLUE_900,
            fill: Colors.ACTION_BLUE_900,
            '&:hover': {
                color: Colors.ACTION_BLUE_1000,
                fill: Colors.ACTION_BLUE_1000
            }
        },
        danger: dangerStyles
    }
});

const invertedVariantStyles = variant({
    variants: {
        default: {
            color: Colors.WHITE,
            fill: Colors.WHITE,
            '&:hover': {
                color: Colors.AUTHENTIC_BLUE_350,
                fill: Colors.AUTHENTIC_BLUE_350
            }
        },
        danger: dangerStyles
    }
});

const TextButton: React.FC<TextButtonProps> = styled(BaseButton)<TextButtonProps>`
    transition: color 200ms;

    ${p => (p.inverted ? invertedVariantStyles(p) : variantStyles(p))};

    &:disabled {
        color: ${p => (p.inverted ? Colors.AUTHENTIC_BLUE_550 : Colors.AUTHENTIC_BLUE_200)};
        fill: ${p => (p.inverted ? Colors.AUTHENTIC_BLUE_550 : Colors.AUTHENTIC_BLUE_200)};
    }
`;

TextButton.defaultProps = {
    size: 'medium',
    variant: 'default'
};

export { TextButton, TextButtonProps };
