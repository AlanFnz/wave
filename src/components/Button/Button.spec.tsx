import React from 'react';
import { render } from '@testing-library/react';
import { Button } from './Button';
import { Colors } from '../../essentials';

describe('standard button', () => {
    it('renders by default the primary variant', () => {
        const { container: primaryTree } = render(<Button variant="primary">Click</Button>);
        const { container: defaultTree } = render(<Button>Click</Button>);

        expect(primaryTree).toEqual(defaultTree);
    });

    describe('primary', () => {
        it('renders as design system suggests', () => {
            const { container } = render(<Button variant="primary">Click</Button>);
            expect(container.firstChild).toHaveStyle(`
              background: ${Colors.AUTHENTIC_BLUE_900};
              color: ${Colors.WHITE};
            `);
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe('secondary', () => {
        it('renders as design system suggests', () => {
            const { container } = render(<Button variant="secondary">Click</Button>);
            expect(container.firstChild).toHaveStyle(`
              background: ${Colors.WHITE};
              color: ${Colors.AUTHENTIC_BLUE_900};
              border-color: ${Colors.AUTHENTIC_BLUE_200};
            `);
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe('danger', () => {
        it('renders as design system suggests', () => {
            const { container } = render(<Button variant="danger">Click</Button>);
            expect(container.firstChild).toHaveStyle(`
              background: ${Colors.NEGATIVE_ORANGE_900};
              color: ${Colors.WHITE};
            `);
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});

describe('inverted button', () => {
    it('renders by default the primary variant', () => {
        const { container: primaryTree } = render(
            <Button inverted variant="primary">
                Click
            </Button>
        );
        const { container: defaultTree } = render(<Button inverted>Click</Button>);

        expect(primaryTree).toEqual(defaultTree);
    });

    describe('primary', () => {
        it('renders as design system suggests', () => {
            const { container } = render(
                <Button variant="primary" inverted>
                    Click
                </Button>
            );
            expect(container.firstChild).toHaveStyle(`
              background: ${Colors.WHITE};
              color: ${Colors.ACTION_BLUE_900};
            `);
            expect(container.firstChild).toMatchSnapshot();
        });
    });

    describe('secondary', () => {
        it('renders as design system suggests', () => {
            const { container } = render(
                <Button variant="secondary" inverted>
                    Click
                </Button>
            );

            expect(container.firstChild).toHaveStyle(`
              background: transparent;
              color: ${Colors.WHITE};
              border-color: ${Colors.WHITE};
            `);
            expect(container.firstChild).toMatchSnapshot();
        });
    });
});
