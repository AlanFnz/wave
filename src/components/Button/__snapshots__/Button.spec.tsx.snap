// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inverted button primary renders as design system suggests 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 0.25rem;
  border: 0.0625rem solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: "Open Sans",sans-serif;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  height: 3rem;
  min-width: 6.25rem;
  padding: 0 1rem;
  -webkit-transition: background ease 216ms,border-color ease 216ms;
  transition: background ease 216ms,border-color ease 216ms;
  background: #FFFFFF;
  border-color: #FFFFFF;
  color: #096BDB;
  fill: #096BDB;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 svg path {
  fill: inherit;
}

.c0:hover {
  background: #096BDB;
  border-color: #096BDB;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.c0:disabled {
  background: #637689;
  border-color: #637689;
  color: #9CA7B4;
  fill: #9CA7B4;
}

<button
  class="sc-AxjAm c0"
>
  Click
</button>
`;

exports[`inverted button secondary renders as design system suggests 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 0.25rem;
  border: 0.0625rem solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: "Open Sans",sans-serif;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  height: 3rem;
  min-width: 6.25rem;
  padding: 0 1rem;
  -webkit-transition: background ease 216ms,border-color ease 216ms;
  transition: background ease 216ms,border-color ease 216ms;
  background: transparent;
  border-color: #FFFFFF;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 svg path {
  fill: inherit;
}

.c0:hover {
  background: #FFFFFF;
  color: #001E3E;
  fill: #001E3E;
}

.c0:disabled {
  background: transparent;
  border-color: #637689;
  color: #637689;
  fill: #637689;
}

<button
  class="sc-AxjAm c0"
>
  Click
</button>
`;

exports[`standard button danger renders as design system suggests 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 0.25rem;
  border: 0.0625rem solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: "Open Sans",sans-serif;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  height: 3rem;
  min-width: 6.25rem;
  padding: 0 1rem;
  -webkit-transition: background ease 216ms,border-color ease 216ms;
  transition: background ease 216ms,border-color ease 216ms;
  background: #FF5E05;
  border-color: #FF5E05;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 svg path {
  fill: inherit;
}

.c0:hover {
  background: #BF4704;
  border-color: #BF4704;
}

.c0:disabled {
  color: #FFFFFF;
  fill: #FFFFFF;
  background: #C6CDD4;
  border-color: #C6CDD4;
}

<button
  class="sc-AxjAm c0"
>
  Click
</button>
`;

exports[`standard button primary renders as design system suggests 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 0.25rem;
  border: 0.0625rem solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: "Open Sans",sans-serif;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  height: 3rem;
  min-width: 6.25rem;
  padding: 0 1rem;
  -webkit-transition: background ease 216ms,border-color ease 216ms;
  transition: background ease 216ms,border-color ease 216ms;
  background: #001E3E;
  border-color: #001E3E;
  color: #FFFFFF;
  fill: #FFFFFF;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 svg path {
  fill: inherit;
}

.c0:hover {
  background: #000F1F;
  border-color: #000F1F;
}

.c0:disabled {
  color: #FFFFFF;
  fill: #FFFFFF;
  background: #C6CDD4;
  border-color: #C6CDD4;
}

<button
  class="sc-AxjAm c0"
>
  Click
</button>
`;

exports[`standard button secondary renders as design system suggests 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: transparent;
  border-radius: 0.25rem;
  border: 0.0625rem solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: "Open Sans",sans-serif;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  height: 3rem;
  min-width: 6.25rem;
  padding: 0 1rem;
  -webkit-transition: background ease 216ms,border-color ease 216ms;
  transition: background ease 216ms,border-color ease 216ms;
  background: #FFFFFF;
  border-color: #C6CDD4;
  color: #001E3E;
  fill: #001E3E;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0 svg path {
  fill: inherit;
}

.c0:hover {
  background: #F1F2F4;
}

.c0:disabled {
  color: #C6CDD4;
  fill: #C6CDD4;
  background: #FFFFFF;
  border-color: #C6CDD4;
}

<button
  class="sc-AxjAm c0"
>
  Click
</button>
`;
